tools/plugins/network/src/main/kotlin/net/corda/cli/plugins/network/BaseOnboard.kt

```
@@ -263,137 +259,177 @@ abstract class BaseOnboard : Runnable {
         }
-        val id = body.get("registrationId").toString()
-        println("Registration ID of $x500Name is $id")
+
+        println("Registration ID of $x500Name is $registrationId")
+
         if (waitForFinalStatus) {
-            waitForFinalStatus(id)
+            waitForFinalStatus(registrationId)
         }
     }
 
-    private fun waitForFinalStatus(id: String) {
-        val end = System.currentTimeMillis() + 5 * 60 * 1000
-        while (System.currentTimeMillis() < end) {
-            Thread.sleep(400)
-            val status = Unirest.get("/membership/$holdingId/$id").asJson()
-            val registrationStatus = status.bodyOrThrow().`object`.get("registrationStatus")
-            when (registrationStatus) {
-                "APPROVED" -> {
-                    return
-                }
-                "DECLINED", "INVALID", "FAILED" -> {
-                    throw OnboardException("Status of registration is $registrationStatus.")
-                }
-                else -> {
-                    println("Status of $x500Name registration is $registrationStatus")
+    private fun waitForFinalStatus(registrationId: String) {
+        createRestClient(MemberRegistrationRestResource::class).use { client ->
+            executeWithRetry(WAIT_DURATION, "Check Registration Progress") {
+                val status = client.start().proxy.checkSpecificRegistrationProgress(holdingId, registrationId)
+
+                when (val registrationStatus = status.registrationStatus) {
+                    RegistrationStatus.APPROVED -> return@executeWithRetry
+                    RegistrationStatus.DECLINED,
+                    RegistrationStatus.INVALID,
+                    RegistrationStatus.FAILED -> throw OnboardException("Status of registration is $registrationStatus.")
+
+                    else -> println("Status of registration is $registrationStatus")
```

Same as my previous comment - we will need to throw an exception in the else case to trigger a retry

---

Will address it

---

I don't think this is a very elegant way to achieve what we want here, i.e. throwing an error in a scenario that's not really exceptional just to let that exception be caught by `executeWithRetry` and get retried. Furthermore, looking at the code of `executeWithRetry` if you do this, you will end up with an error log of the form "Unable to perform operation" even though the operation was performed fine. 

Here are a few ways I can think we can achieve this in a better way (open to other suggestions):
* Adjust the `executeWithRetry` to accept a lambda that will be used to determine if the result of the operation is what was expected, in which case it can be returned, or otherwise continue retrying with an appropriate log message (different to the one shown when the operation fails).
* Create a separate utility function that does only this looping without caring about any exceptions (as these will be caught by `executeWithRetry`).

The second one might be a bit simpler, judging from the code in `executeWithRetry` at the moment, but trying to write the code for each approach will probably give a better idea of what's better and easier to maintain.

---

Thinking about it, I am not even sure `executeWithRetry` is necessary at the moment given the network plugin doesn't have a strong need to be tolerant against individual request failures. We can add that if we want, but not critical for this work it can be done separately. 

I see the original comment [here](https://github.com/corda/corda-runtime-os/pull/4312#discussion_r1266822570) was around not implementing retry logic ourselves. But, `executeWithRetry` is designed to handle only exceptions at the moment. So, just implementing a utility function that does that continuous checking of an invariant might be much simpler really (just a while loop with a sleep).